### ISSUE TITLE
Migrate to version 4 of Mattermost API

### ISSUE THREAD

**Issue by jtdaugherty** [matterhorn-chat/matterhorn#303]
One path is outlined in #202.

I asked the developers:

I'm going to be looking at upgrading Matterhorn's API use from v3 to v4 over the next few months, and that made me wonder: when switching to API v4 from v3, is it just endpoints that changed? Or are any fundamental response data structures that changing, too? E.g., does a channel or a user object look the same in V4? Is there a migration guide I can look at?

Their response:

Ok, so for the most part the main changes from v3 to v4:
* Endpoint URLs changing to be more consistent and remove the requirement on team IDs
Due to legacy architecture, v3 required a team ID for many endpoints that didn't really require it. For example, getting a channel by ID in v3 required a team and channel ID in the URL, while in v4 it only requires the channel ID since is globally unique
* Collections for the most part now return lists, except for some special cases mentioned in the reference (for example getting posts will return a custom JSON object)
* A lot of the user endpoints in v3 performed actions based on the user ID in the logged in session. In v4 the user ID is required explicitly in almost all of the user endpoints so that admins can have more power
Paging in v3 was done using offset/limit in the URL path, in v4 it is now done using the page/per_page query parameters

Other than those, there are some small changes here and there but if you use the documentation at https://api.mattermost.com they should all be documented

There's also a config setting for disabling v3 endpoints on your server which can be helpful for detecting apps/integrations trying to use v3 endpoints. https://api.mattermost.com/#tag/APIv3-Deprecation

**Comment by jtdaugherty** [comment]
See also https://api.mattermost.com/#tag/APIv3-Deprecation

**Comment by jtdaugherty** [comment]
This ticket supersedes the older #202.

**Comment by jtdaugherty** [comment]
@jasonblais See also this ticket.

**Comment by jtdaugherty** [comment]
@aisamanra is the one doing the work to migrate `mattermost-api` to v4.

**Comment by jasonblais** [comment]
Perfect, great to hear! 👍

**Comment by aisamanra** [comment]
Current status on this is that the relevant bindings have been merged into `develop` and the current branch uses exclusively version 4 of the API. On the `mattermost-api` library side, there are a few more things to do: for one, all the old endpoints need deprecation warnings before the next release, and the export structure needs a bit of cleaning up, and importantly we should generate lenses for all the new types.

It's probably worth noting in this issue that the `mattermost-api` bindings for version 4 of the API were based on code generated from the [Mattermost Swagger specification](https://api.mattermost.com/static/mattermost-openapi-v4.yaml). This involved some heuristics, and consequently the generated code wasn't used directly. Several more endpoints are generated but not yet exposed, but for most of them it should be a matter of uncommenting and then making sure that they behave in the correct way.

(An example of changes needed is that Swagger doesn't consistently report whether certain JSON fields are necessarily optional: sometimes, we'd generate a Haskell type from the Swagger spec and only after trying to decode it would we discover that a given field could be missing, which would require mild human intervention into the generated code. For the most part, though, it let us do a ton of work without a ton of effort.)

**Comment by jtdaugherty** [comment]
This is now done and released in `40600.0.0`.

**Comment by jasonblais** [comment]
Great! Thanks @jtdaugherty!
